src/main/java/htsjdk/variant/variantcontext/writer/VCFWriter.java

```diff
@@ -223,9 +231,21 @@ public void add(final VariantContext context) {
             write("\n");
 
             writeAndResetBuffer();
-
+            isBodyWritten = true;
         } catch (IOException e) {
             throw new RuntimeIOException("Unable to write the VCF object to " + getStreamName(), e);
         }
     }
+
+    @Override
+    public void setVcfHeader(VCFHeader header) {
+        if (isHeaderWritten) {
+            throw new IllegalStateException("Header is already written");
+        }
+        if (isBodyWritten) {
+            throw new IllegalStateException("Body is already written");
+        }
```

Comment:
Same comments as above about using a single state variable, more specific error message, and final args.